docs/tpm2-pkcs11_object_auth_model.md

```diff
@@ -0,0 +1,528 @@
+# tpm2-pkcs11 object & authorization model
+
+## Introduction
+
+The objective of this document is to present an architectural overview of the
+tpm2-pkcs11 object and authorization model implemented in the design. The topics
+covered here in are:
+
+1. [PKCS11 Objects](#pkcs11-objects)
+2. [PKCS11 Roles](#pkcs11-roles)
+3. [TPM2 Objects](#tpm2-objects)
+4. [tpm2-pkcs11 Objects](#tpm2-pkcs11-objects)
+5. [NV Object Diagram](#nv-object-diagram)
+6. [NV Policy Diagram](#nv-policy-diagram)
+7. [Transient Object Diagram](#transient-object-diagram)
+8. [Transient Policy Diagram](#transient-policy-diagram)
+9. [Sample tpm2-tools implementation](#sample-tpm2-tools-implementation)
+10. [Important note on authorization](#important-note-on-authorization)
+
+## PKCS11 Objects
+An Overview of the hierarchical concept of objects.
+<p>
+  <img src="illustrations/reader-slot-token-obj.png"
+</p>
+<br>
+<br>
+
+* The PKCS11 READER is a medium through with information is exchanged with the
+PKCS11 device or TOKEN.
+* Every READER has a SLOT that may contain a TOKEN.
+* The TOKEN ultimately protects all the USER created objects as well as access
+through an authorization model to distinguish the SO, USER or adversary.
+* USER objects can be keys, certificates or any user defined data.
+
+## PKCS11 Roles
+Overview of pkcs11 requirements of the security officer and user roles. The
+PKCS11 design describes two specific roles - A security-officer or **SO**
+and **USER**. These roles are realized with separate PIN authentications.
+Following are some relevant design considerations when comparing the roles:
+* Neither SO nor USER can know each other's PIN.
+* The SO is restricted to an administrative role to modify SO PIN and or an
+existing USER's PIN.
+* While both the **SO and the USER can change USER PIN**, only a SO can change
+SO PIN.
+* USER authentication PIN is a must for operating on USER objects.
+* The SO cannot access any USER objects.
+* While SO could change USER PIN, the **SO cannot restore the old USER PIN**.
+This intentionally prevents SO from having explicit access to USER objects and
+also to bring traceability to security officer actions regarding PIN changes for
+USER authorization.
+<br>
+As an example, a token initialized with a SO PIN can be setup with a default
+USER PIN. The USER then changes the PIN to his liking which the SO cannot know
+& thus cant restore. The USER can now use the new PIN to create and operate on
+USER objects. If the USER is convinced of forgetting the PIN, USER can get the
+PIN reset by the SO. However, if the SO resets the USER PIN without USER
+request, the old USER PIN cannot be restored or used; thus revealing the SO act
+of changing the USER PIN.
+
```

Comment:
>    tpm2-pkcs11 currently uses PBKDF2, but (1) it uses it for off-TPM authenticated encryption, which provides a PIN-testing oracle that can be used to circumvent TPM mitigations (see #278); and (2) PBKDF2 is unnecessary when TPM mitigations are correctly used, and it just slows down legitimate usage.

The TPM itself slows down legitimate usage. Speed really isn't an issue here because the TPM is so slow. The PBKDF when you don't have access to the salt strengthens weak pins.

Reply:
> The TPM itself slows down legitimate usage.

Using the TPM is slower than not using the TPM, but using the TPM is also more secure than not using the TPM. That's a reasonable trade off for some use cases.

However, using PBKDF2 is slowing than not using PBKDF2, but (assuming TPM w/ DA protection) isn't any more secure (see below). That's a bad trade off.

> Speed really isn't an issue here because the TPM is so slow.

Disagree. The TPM is already slow, so we shouldn't slow its usage down further.

> The PBKDF when you don't have access to the salt strengthens weak pins.

RFC 8018, section 4.1 explains that using a salt has two benefits, neither of which are relevant to the use case of TPM2 auth values:

```
      1.  It is difficult for an opponent to precompute all the keys, or
          even the most likely keys, corresponding to a dictionary of
          passwords.  If the salt is 64 bits long, for instance, there
          will be as many as 2^64 keys for each password.  An opponent
          is thus limited to searching for passwords after a password-
          based operation has been performed and the salt is known.
```

This benefit is about preventing rainbow table attacks on password-hash databases. That is, the goal is to prevent an attacker from being able to common password hashes up front and then look for collisions with a leaked password-hash database.

Precomputing keys is not a threat to TPM2 auth values, because the attacker doesn't have access to the TPM2 auth value itself. The attacker has to sequentially feed computed keys to the TPM anyway.

```
      2.  It is unlikely that the same key will be selected twice.
          Again, if the salt is 64 bits long, the chance of "collision"
          between keys does not become significant until about 2^32 keys
          have been produced, according to the Birthday Paradox.  The
          fact that collisions are unlikely addresses some concerns
          about interactions between multiple uses of the same key that
          may arise when using some encryption and authentication
          techniques.
```

This benefit is about ensuring that encryption/authentication schemes that aren't safe under key reuse (e.g., CTR mode for block ciphers) don't have to worry if the user reuses the same password.

Again, this isn't relevant to TPM2 auth values. Auth values are simply compared for equality; two TPM2 objects that use the same auth value aren't any weaker than two TPM2 objects that use different auth values.

The aspect of PBKDF2 that makes passwords stronger is not the salt, but the iteration count. As RFC 8018, section 4.2 says:

```
    An iteration count has traditionally served the purpose of increasing
    the cost of producing keys from a password, thereby also increasing
    the difficulty of attack.
```

The iteration count accomplishes this by slowing down the attacker's ability to guess passwords... but this is exactly what the TPM's DA protection does.

--

I'll agree that a **secret** salt (that is, unknown to the attacker) does make guessing the auth value harder... but then you're really talking about a *key* (something that must be kept secret to provide security) rather than a *salt* (something that is okay for an attacker to know without compromising security).

And in that case, rather than using PBKDF2, you can just *concatenate* this secret with the user's PIN (and hash once with NameAlg, if necessary for size reasons) to form the TPM auth value.

Reply:
> 
> ```
>       1.  It is difficult for an opponent to precompute all the keys, or
>           even the most likely keys, corresponding to a dictionary of
>           passwords.  If the salt is 64 bits long, for instance, there
>           will be as many as 2^64 keys for each password.  An opponent
>           is thus limited to searching for passwords after a password-
>           based operation has been performed and the salt is known.
> ```

That's the important part "and the salt is known". When the salt isn't known its a secret mixed in, when the salt is known it helps to prevent rainbow tables. This requires an attacker to both get the pin and the salt correct. The iterations really do nothing for us, we could really do:

namealghash(pin + random[64]) and get those properties.